### PR TITLE
Implemented list projects for GitServices (#485 )

### PR DESCRIPTION
**Pull request by abkosar**
Implemented list projects for GitServices (#485 )

@lachmanfrantisek 
@csomh

### CODE REVIEW

File: ogr/services/github/service.py
Comment (lachmanfrantisek) [inline ogr/services/github/service.py:197]: Just to check, will an empty space at the beginning of `search_query` work well in case of `user=None and langure="something"`?

**Review by lachmanfrantisek — COMMENTED** [review]
Core looks good to me, thanks!

Can you please add some test cases for this functionality? Get inspired by other tests, check our [contribution guide](https://github.com/packit/ogr/blob/master/CONTRIBUTING.md#testing). And let us know if you need some help with that.

**Comment by abkosar** [comment]
Sounds good, will definitely do that!

File: ogr/services/github/service.py
Comment (abkosar) [reply ogr/services/github/service.py:197]: Yeah good question. I will take a look at that but I don't think it's going to be a problem.

**Comment by lachmanfrantisek** [comment]
@abkosar How you are doing? Do you need some help?

**Comment by abkosar** [comment]
@lachmanfrantisek I'm doing good. I haven't forgotten about this but work has been very busy for the past month. I started working through a google cloud certification for machine learning so it has been taking a lot of my time lately but I would like to finish this if that's ok.

But before this certification started, I looked at the tests. I haven't used tox.ini before so I will need some guidance. You said to add some test cases. My guess would be to add those cases to the `.yaml` files since the test scripts are already written?

**Comment by lachmanfrantisek** [comment]
Sure, that's ok. Good luck with the certification.

Regarding the tests, you can run tests via `pytest` directly, or use `make check`. (See [our contribution guide](https://github.com/packit/ogr/blob/main/CONTRIBUTING.md#testing).


> My guess would be to add those cases to the .yaml files since the test scripts are already written?

You don't need to do this manually. Just write some test case with real data and [requre](https://github.com/packit/requre) will save those responses for you. (See the guide and let us know if you hit some problem with that approach.)

**Comment by abkosar** [comment]
@lachmanfrantisek 

So I tried to do `git fetch --all && git rebase upstream/main` and I got the following message:

```
On branch master
Your branch and 'origin/master' have diverged,
and have 61 and 10 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        .idea/

nothing added to commit but untracked files present (use "git add" to track)

```

This diverge issue happened to me before doing the pull request and I solved it by spending a lot of time. I remember it asked me to run through all my commits again. I just wanted to see if there is an easier way to handle this.

Thanks!

**Comment by TomasTomecek** [comment]
@abkosar that's actually expected since our upstream repository no longer has a "master" branch.

> On branch master

Please run the same rebase command in your local branch for this PR and force-push the changes - that should resolve the conflicts (which you may need to manually).

**Comment by abkosar** [comment]
@TomasTomecek thanks for the advice I got rid of that error, moved on to writing tests. 

I'm using PyCharm as my IDE and there is this `.idea/` folder. In the `.gitignore` file the files in my `.idea/` folder isn't included. Can I just update the .gitignore file?

**Comment by TomasTomecek** [comment]
> @TomasTomecek thanks for the advice I got rid of that message, moved on to writing tests.
> 
> I'm using PyCharm as my IDE and there is this `.idea/` folder. In the `.gitignore` file, the files in my `.idea/` folder isn't included. Can I just update the .gitignore file?

Well, no offence, but we are not interested in your .idea folder :) that's your local configuration and it's not beneficial to share (the same way you are probably not interested in our .idea config files since some of us use pycharm as well)

**Comment by abkosar** [comment]
haha none taken:) Good to know, thanks.

**Comment by abkosar** [comment]
@TomasTomecek So I wrote some tests for the `list_projects` method I implemented for `github` and `gitlab` services. While I was implementing the tests I figured some improvements in the function (I guess this is the whole point of writing tests :) ) so I am also updating them. 

I'm working on a mac and when I tried to run `make prepare-check && make check` or `make build` it wasn't successful because with either one I didn't have something installed which I guess meant for Linux systems?

Then I tried running `pytest tests/integration/gitlab/test_service.py` and `pytest tests/integration/github/test_service.py` and those passed just fine.

Is the correct way to do it?

Thanks!

**Comment by TomasTomecek** [comment]
@abkosar yes, sounds correct if you were able to run the tests using pytest locally; in the end, we want to be sure the tests pass in our CI, that will be the ultimate proof

**Comment by abkosar** [comment]
I already see somethings I missed about the pre-commit hooks. I will fix those and squash the commits too.

File: tests/integration/github/test_service.py
Comment (TomasTomecek) [inline tests/integration/github/test_service.py:90]:
```suggestion
        projects = self.service.list_projects(user=user)
```
you should be able to replace this since service is defined for the whole test class

File: tests/integration/github/test_service.py
Comment (TomasTomecek) [inline tests/integration/github/test_service.py:98]: same as above
```suggestion
        projects = self.service.list_projects(user=user, language=language)
```

File: tests/integration/gitlab/test_service.py
Comment (TomasTomecek) [inline tests/integration/gitlab/test_service.py:97]:
```suggestion
        projects = self.service.list_projects(namespace=name_of_the_repo)
```

the same here as well :)

**Review by TomasTomecek — COMMENTED** [review]
awesome, you're already getting there! very nice

I can see that we define the `service` object globally for the whole test class, so you can drop those initialization lines

glad to see that you were able to generate the test data

**Comment by abkosar** [comment]
@TomasTomecek I guess this PR looks good?

I know this may have been the longest PR but if that's ok I'd like to continue contributing. This PR originally included implementation of pagure but I first wanted to complete gitlab and github.

So, let me know your thoughts, thanks!

**Review by TomasTomecek — APPROVED** [review]
LGTM, very nice! Asking for another pair of eyes here from other maintainers.

File: ogr/services/gitlab/service.py
Comment (lachmanfrantisek) [inline ogr/services/gitlab/service.py:145]: Can we have a better name, please?

```suggestion
            user_object = self.gitlab_instance.users.list(username=user)[0]
            projects = user_object.projects.list(all=True)
```

File: ogr/services/gitlab/service.py
Comment (lachmanfrantisek) [inline ogr/services/gitlab/service.py:178]: This if statement is pretty huge and the core part is duplicated.

What about something like this?


```suggestion
        if language:
            # group.projects.list gives us a GroupProject instance
            # in order to be able to filter by language we need Project instance
            projects_to_convert = [
                self.gitlab_instance.projects.get(item.attributes["id"])
                for item in projects
                if language
                in project.languages().keys()
            ]
        else:
            projects_to_convert = projects

        gitlab_projects = [
            GitlabProject(
                repo=project.attributes["path"],
                namespace=project.attributes["namespace"]["full_path"],
                gitlab_repo=project,
                service=self,
            )
            for project in projects_to_convert
        ]
```

**Review by lachmanfrantisek — COMMENTED** [review]
Thanks for all the work, I have only two small style improvements.

File: ogr/services/gitlab/service.py
Comment (mfocko) [inline ogr/services/gitlab/service.py:143]: I would probably use `elif` here and in case of setting both `namespace` and `user` raise an exception, we should have `OperationNotSupported`

File: ogr/services/pagure/service.py
Comment (mfocko) [inline ogr/services/pagure/service.py:299]: as mentioned above, `ogr/exceptions.py` should have exception class for not supported operations :)

**Review by mfocko — COMMENTED** [review]
otherwise LGTM, thanks for your contribution :)

File: ogr/services/gitlab/service.py
Comment (abkosar) [reply ogr/services/gitlab/service.py:178]: Yeah this totally makes sense. I will update the code accordingly.

File: ogr/services/gitlab/service.py
Comment (abkosar) [reply ogr/services/gitlab/service.py:145]: You're right. I forgot about the variable names, you told me this one before, my bad.

File: ogr/services/gitlab/service.py
Comment (abkosar) [reply ogr/services/gitlab/service.py:143]: Thanks for clearing this. I actually thought about how to handle this but couldn't come up with a way. I will update the code accordingly, thanks!

**Comment by abkosar** [comment]
There was this weird error going on when I ran the tests I wrote locally:
```
================================================================================================================================= short test summary info ==================================================================================================================================
FAILED tests/integration/gitlab/test_service.py::Service::test_list_projects_with_namespace_language_input - requre.exceptions.StorageNoResponseLeft: ('No responses left. Try to regenerate response file (/Users/ardkosar1/Documents/ogr/tests/integration/gitlab/test_data/test_servic...
======================================================================================================================== 1 failed, 5 deselected, 1 warning in 1.27s ========================================================================================================================
```

I fixed it by changing the name of the second test. But I have no idea why it fixed this but should be good now.

**Comment by TomasTomecek** [comment]
> There was this weird error going on when I ran the tests I wrote locally:
> 
> ```
> ================================================================================================================================= short test summary info ==================================================================================================================================
> FAILED tests/integration/gitlab/test_service.py::Service::test_list_projects_with_namespace_language_input - requre.exceptions.StorageNoResponseLeft: ('No responses left. Try to regenerate response file (/Users/ardkosar1/Documents/ogr/tests/integration/gitlab/test_data/test_servic...
> ======================================================================================================================== 1 failed, 5 deselected, 1 warning in 1.27s ========================================================================================================================
> ```
> 
> I fixed it by changing the name of the second test. But I have no idea why it fixed this but should be good now.

That's actually what the error message is trying to tell - we should improve it so it's more clear, filing an issue for that: https://github.com/packit/requre/issues/191

@lachmanfrantisek @mfocko are we good here, is this ready for a merge?

@abkosar you are a true champion! Thank you for sticking in and keeping up with our comments.

**Comment by mfocko** [comment]
> @mfocko are we good here, is this ready for a merge?

looks OK, only rebase, but we can do it :)

File: tests/integration/gitlab/test_service.py
Comment (lachmanfrantisek) [inline tests/integration/gitlab/test_service.py:97]: Since you've renamed the test, the response file `tlab/test_data/test_service/Service.test_list_projects_with_namespace_language_input.yaml` is not needed and can be removed.

**Review by lachmanfrantisek — APPROVED** [review]
Well done! :rocket:

**Comment by abkosar** [comment]
@TomasTomecek Thanks! Well I am here to learn and I actually learned a lot just from this PR. Thanks for waiting for me to complete it!

**Comment by mfocko** [comment]
rebasing

**Comment by TomasTomecek** [comment]
Whaaaaat, the rebase actually closed the PR? That's unfortunate.

**Comment by mfocko** [comment]
@abkosar Just out of interest, did you uncheck something like _maintainers are allowed to push_ in the PR? Or it's just GitHub mechanism for protecting `main`

**Comment by mfocko** [comment]
I'll reopen the PR from my branch, credit is all yours, you are preserved as author of the commits :)

**Comment by abkosar** [comment]
@mfocko No I didn't. I don't even know how to uncheck that. But maybe it's the default setting? I didn't do anything on purpose though.

**Comment by mfocko** [comment]
No worries :) it's fine, must be GitHub then :smile: 

will be merged via #588

**Pull request closed without merging**
